### PR TITLE
Userclass

### PR DESCRIPTION
**Pull request by Chriscastanuela**
## Description
Clearly and concisely describe the feature.
* Completes the functionality of the original project's rubric

## Where To Start
* Review the methods in the classes

## Affected areas of application
* This creates functionality but we still need to edit the DOM manipulation in order to use said functionality

## Screenshots (if applicable):
* N/A

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] Test A - Chai and Mocha
- [x] Test B - Console logging

## Relevant Tickets
* https://github.com/Chriscastanuela/whats-cookin/projects/1#card-44952439
* https://github.com/Chriscastanuela/whats-cookin/projects/1#card-45117186
## Tags
* N/A
## Questions or Notes (if applicable):
* Commit 0b13b34 fixes a pantry method

### CODE REVIEW

File: src/pantry.js [hunk 1]
```diff
@@ -28,10 +28,10 @@ class Pantry {
       return index.ingredient;
     })
     let notInThere = [];
-    let ings = recipeData.filter(index => {
+    let ings = recipeData.find(index => {
       return index.id  == nameOrId || index.name == nameOrId;
-    }).ingredients;
-    ings.forEach(i => {
+    });
+    ings.ingredients.forEach(i => {
```
Comment (holladayian) [inline src/pantry.js]: We should refactor the name of "ings" to something more semantic. Possibly, "ingredientsFromRecipie"?

**Review by holladayian — APPROVED** [review]
Just a minor thought, but we should review naming techniques.

File: src/user.js [hunk 2]
```diff
@@ -77,14 +79,36 @@ class User {
         var newObject = {
         name: dex.name,
         id: dex.id,
-        hasEnough: "You don't have any of this item"
+        hasEnough: -dex.quantity.amount
```
Comment (holladayian) [inline src/user.js:82]: I know this works, and it looks good to me, but I also know we were wondering if this is the best technique in regards to returning a negative property "-dex.quantity.amount". It we have the time, let's ask a mentor.

File: src/user.js [hunk 3]
```diff
@@ -77,14 +79,36 @@ class User {
...
-    console.log(newArray)
     return newArray;
   }
+
+  returnShoppingList(recipeIngredients) {
+    let groceries = this.returnAmount(recipeIngredients);
+    let notEnoughGroceries = groceries.filter(index => {
+      return index.hasEnough < 0
+    })
+    let foundIt = [];
```
Comment (holladayian) [inline src/user.js:95]: Maybe we can change "foundIt" to "foundIngredientList".

File: src/user.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ class User {
     this.name = name;
     this.pantry = pantry;
     this.favoriteRecipes = [];
-
+    this.recipesToCook = [];
```
Comment (bretm9) [inline src/user.js:9]: This is a necessary addition for when we implement cookRecipe. Looks good.

File: test/user-test.js [hunk 5]
```diff
@@ -171,20 +217,23 @@ describe('User', () => {
...
-    { name: 'fresh basil', id: 2044, quantity: [Object] },
-    { name: 'grape tomatoes', id: 10111529, quantity: [Object] },
-    { name: 'olive oil', id: 4053, quantity: [Object] },
-    { name: 'zucchini', id: 11477, quantity: [Object] }
-    */
-    expect(user1.returnAmount(recipeOne.ingredients)).to.eql(recipeData[47]);
-
+  it('12. Should check quantities for each ingredient', () => {
+    
+    expect(user2.returnAmount(recipeOne.ingredients)).to.eql(
```
Comment (holladayian) [inline test/user-test.js:222]: If we really wanted to cut lines out we could ```expect(user2.returnAmount(recipeOne.ingredients)[0])``` and then use only line 224 in regards to what it should be equal to. That's only if we're going to make this entire project 27 lines total 😎

**Review by holladayian — APPROVED** [review]
Minor clean-up, naming techniques and a question to a mentor/authority about math functions.

File: src/user.js [hunk 6]
```diff
@@ -7,25 +7,25 @@ class User {
...
-    return this.favoriteRecipes.filter(recipe => {
+  filterRecipes(tag, property) {
+    return this[property].filter(recipe => {
       return recipe.tags.includes(tag);
     });
   }
 
-  findFavorites(strgToSrch) {
-    return this.favoriteRecipes.filter(recipe => {
+  findInCategory(strgToSrch, property) {
```
Comment (bretm9) [inline src/user.js:29]: Passing in properties here will let us search through the appropriate category (favoriteRecipes or recipesToCook). We'll see this in action on the DOM soon

File: src/user.js [hunk 7]
```diff
@@ -69,11 +67,11 @@ class User {
     let newArray = [];
     recipeIngredients.forEach(dex => {
       if (pantryIds.includes(dex.id)) {
+          let theIndex = pantryIds.indexOf(dex.id);
           var newObject = {
           name: dex.name,
           id: dex.id,
-          // hasEnough: dex.quantity.amount -= dex.quantity.amount
-          hasEnough: "You got what it takes"
+          hasEnough: this.pantry[theIndex].amount -= dex.quantity.amount
```
Comment (holladayian) [inline src/user.js:76]: We might be able to cahnge "hasEnough" to something more semantic, like "leftoverIngredientAmount".

**Review by holladayian — APPROVED** [review]
Just a bit more small naming techniques.

File: src/pantry.js [hunk 8]
```diff
@@ -1,7 +1,28 @@
+import recipeData from "./data/recipes";
+
 class Pantry {
   constructor(userIngredients) {
     this.contents = userIngredients;
   }
+  doIHaveEnoughForThis(nameOrId) {
```
Comment (bretm9) [inline src/pantry.js:7]: This method could be renamed to `checkForIngredients` but this name is very explanatory which I like.

File: src/pantry.js [hunk 9]
```diff
@@ -1,7 +1,28 @@
...
+    let toggle = "Yes";
```
Comment (bretm9) [inline src/pantry.js:8]: This toggle variable could be refactored to be assigned to true or false to work nicely in conditionals

File: src/user.js [hunk 10]
```diff
@@ -64,77 +64,53 @@ class User {
...
-  //         }
-  //     })
-  //     newArray.push(newObject);
-  //   })
-  //   console.log(newArray);
-  //   return newArray;
-  // }
+  returnAmount(recipeIngredients) {
+    let pantryIds = this.pantry.map(index => index.ingredient);
+    let newArray = [];
```
Comment (holladayian) [inline src/user.js:69]: Let's not forget, we shouldn't name things using their data type. I also don't know if we should save "new" for new class instantiations. I think ```newArray``` can be could be changed to ```recipeIngredientsFromPantry```

File: src/user.js [hunk 11]
```diff
@@ -64,77 +64,53 @@ class User {
...
+          hasEnough: "You got what it takes"
+        }
+      } else {
+        var newObject = {
+        name: dex.name,
+        id: dex.id,
+        hasEnough: "You don't have any of this item"
+        }
+      }
+      newArray.push(newObject);
```
Comment (holladayian) [inline src/user.js:85]: Along with the naming technique mentioned above, we could rename newObject to something like ```specificRecipeIngredientFromPantry```. I know it's long, and going to be a pain to write it out over and over, but having a well defined name will help us from losing track of it in a big project like this, especially when we are refactoring someone else's work.

File: src/pantry.js [hunk 12]
```diff
@@ -23,6 +22,22 @@ class Pantry {
     });
     return toggle;
   };
+  
+  howManyIngredientsLeft(nameOrId) {
+    let contentsIds = this.contents.map(index => {
+      return index.ingredient;
+    })
+    let notInThere = [];
```
Comment (bretm9) [inline src/pantry.js:30]: notInThere could be renamed to ingredientsNotInPantry for readability.

**Review by holladayian — APPROVED** [review]
More naming technique changes. I think these are more pertinent than the others, as they don't really describe what this value should be. Should be a pretty easy fix though.

File: src/pantry.js (hunk 9 above)
Comment (holladayian) [reply src/pantry.js:8]: What about the superposition; toggle = "maybe"???

File: src/user.js [hunk 13]
```diff
@@ -34,7 +34,20 @@ class User {
   }
 
   checkPantry(recipeIngredients) {
-    //
+    let toggle;
+    toggle = 'You have the ingredients!';
```
Comment (holladayian) [inline src/user.js]: I know we mentioned toggle being reassigned to a t/f value. This is just another place we can use that.

**Review by holladayian — APPROVED** [review]
This review was mostly covering naming techniques. Most everything else looked good!

**Comment by holladayian** [comment]
I think this looks pretty good! Mainly just a few naming techniques and an odd-ended question to a mentor, regarding math functions. Everything else looks pretty good! Good job y'all!!

File: src/user.js [hunk 14]
```diff
@@ -77,14 +79,36 @@ class User {
...
+    })
+    return shoppingList;
+  }
 }
 
-export default User;
\ No newline at end of file
+export default User;
+
+/*
```
Comment (bretm9) [inline src/user.js]: We'll want to remove this pseudocode in the future, but that can be on the back-burner. I'll make an issue for that now.

**Review by bretm9 — APPROVED** [review]
PR looks good to me! Once I see Ian's approval I'll go ahead merge.

**Comment by bretm9** [comment]
Looks like we're all in approval here. I'll merge it on through to Main. 🎉

**Pull request merged**